src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs

```diff
@@ -163,17 +170,39 @@ public static partial class SymbolFinder
             return ImmutableArray<ISymbol>.Empty;
         }
 
+        #region derived classes
+
         /// <summary>
-        /// Finds the derived classes of the given type. Implementations of an interface are not considered "derived", but can be found
-        /// with <see cref="FindImplementationsAsync(ISymbol, Solution, IImmutableSet{Project}, CancellationToken)"/>.
+        /// Finds all the derived classes of the given type. Implementations of an interface are not considered
+        /// "derived", but can be found with <see cref="FindImplementationsAsync(ISymbol, Solution,
+        /// IImmutableSet{Project}, CancellationToken)"/>.
         /// </summary>
         /// <param name="type">The symbol to find derived types of.</param>
         /// <param name="solution">The solution to search in.</param>
         /// <param name="projects">The projects to search. Can be null to search the entire solution.</param>
         /// <param name="cancellationToken"></param>
         /// <returns>The derived types of the symbol. The symbol passed in is not included in this list.</returns>
-        public static async Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
+        public static Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
             INamedTypeSymbol type, Solution solution, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
+        {
+            return FindDerivedClassesAsync(type, solution, transitive: true, projects, cancellationToken);
+        }
+
+        /// <summary>
+        /// Finds the derived classes of the given type. Implementations of an interface are not considered
+        /// "derived", but can be found with <see cref="FindImplementationsAsync(ISymbol, Solution,
+        /// IImmutableSet{Project}, CancellationToken)"/>.
+        /// </summary>
+        /// <param name="type">The symbol to find derived types of.</param>
+        /// <param name="solution">The solution to search in.</param>
+        /// <param name="transitive">If the search should stop at immediately derived classes, or should continue past that.</param>
+        /// <param name="projects">The projects to search. Can be null to search the entire solution.</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>The derived types of the symbol. The symbol passed in is not included in this list.</returns>
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+        public static async Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
+            INamedTypeSymbol type, Solution solution, bool transitive, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
+#pragma warning restore RS0026 // Do not add multiple public overloads with optional parameters
```

Comment:
the new entrypoints add a 'transitive' flag to let the caller determine if they want just immediate links, or the whole tree of links.

these new entrypoings then call into an internal overload that returns an immutable array.  i made the new public entrypoints still return IEnumerable for consistency with teh existing methods.  But we could consider making the new public ones just return ImmutableArray.  That way public callers can also avoid the boxing.  Thoughts?  

@jasonmalinowski i know you ran into this recently with the workspace apis.  What do you think?

Reply:
So, no matter what the allocation of hte large list is happening.  It's not like these are actually lazy.  And i don't really want IEnumerables because it can lead to accidents, and it is more difficult for conusmers to deal with.  i.e. everyone has to assume the worst and convert back to different forms and whatnot.